src/backend/distributed/utils/multi_partitioning_utils.c

```diff
@@ -130,6 +138,107 @@ worker_fix_pre_citus10_partitioned_table_constraint_names(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * fix_partition_shard_index_names fixes the index names of shards of partitions of
+ * partitioned tables on workers.
+ *
+ * When running CREATE INDEX on parent_table, we didn't explicitly create the index on
+ * each partition as well. Postgres created indexes for partitions in the coordinator,
+ * and also in the workers. Actually, Postgres auto-generates index names when auto-creating
+ * indexes on each partition shard of the parent shards. If index name is too long, it
+ * truncates the name and adds _idx postfix to it. However, when truncating the name, the
+ * shardId of the partition shard can be lost. This may result in the same index name used for
+ * the partition shell table and one of the partition shards.
+ * For more details, check issue #4962 https://github.com/citusdata/citus/issues/4962
+ *
+ * fix_partition_shard_index_names renames indexes of shards of partition tables to include
+ * the shardId.
+ */
+Datum
+fix_partition_shard_index_names(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	EnsureCoordinator();
+
+	if (!PartitionedTable(relationId))
+	{
+		ereport(ERROR, (errmsg("could not fix partition shard index names: "
+							   "relation does not exist or is not partitioned")));
+	}
+	if (!IsCitusTable(relationId))
+	{
+		ereport(ERROR, (errmsg("fix_partition_shard_index_names can "
+							   "only be called for distributed partitioned tables")));
+	}
+
+	Relation relation = relation_open(relationId, AccessShareLock);
```

Comment:
I figured an AccessShareLock would be needed on the table in the coordinator as we wouldn't want for the table to be concurrently dropped. Maybe we can do `try_relation_open` as one table can be dropped before reaching that oid during `fix_all_partition_shard_index_names`

Reply:
Update: there was a mixup since `PartitionedTable()` function already tried to open the relation. Now I only try to open the relation once and I have separated the error cases for "not existing" (adding an isolation test) and "not being partitioned table".

Reply:
I liked the new version much more. Thanks!